### ISSUE TITLE
Text rendering blurry not retina.

### ISSUE THREAD

**Issue by dw9694** [onivim/oni#2047]
Faced such a problem.
I have no retina, my display is 1920x1080, the operating system is Linux Ubuntu 16.04.
The screen blurs in the center, but at the top and bottom is normal.
Thank you.

![2018-04-06 10-25-26](https://user-images.githubusercontent.com/6016334/38408748-86749b78-3987-11e8-97fc-dc71a4ccacf5.png)

**Comment by MarcusE1W** [comment]
Hi.

I think I have the same problem.
I have a 
- notebook with 1366x767 resolution.
- Archlinux
- SpectrWM
- Oni 0.3.1

![2018-04-06-125712_1366x768_scrot](https://user-images.githubusercontent.com/26278354/38420029-6a5e96ba-399a-11e8-9556-58ccf9ac31a9.png)

My impression is that the left Eplorer part is ok, only the text window is mushy.

What log would you need ?

**Comment by jordwalke** [comment]
What could possibly be causing this?

**Comment by ArthurZ** [comment]
Hi @CrossR it is related to https://github.com/onivim/oni/issues/1252 and I was told that it is due to issues around the subpixel anti-aliasing @jordwalke

**Comment by MarcusE1W** [comment]
My impression is that `"ui.fontSmoothing": "subpixel-antialiased"` improves the font sharpness. Thanks.

**Comment by dw9694** [comment]
I do not see the difference

![xojo](https://user-images.githubusercontent.com/6016334/38819277-a023fd74-41a3-11e8-894b-66fffed23244.png)

**Comment by bryphe** [comment]
Thanks for the updates, @MarcusE1W and @dw9694 !

Kind of makes sense, as looking at the screenshots, @MarcusE1W 's screenshot definitely did not have subpixel anti-aliasing enabled. Unfortunately, @dw9694 , yours _does_ have subpixel anti-aliasing, so it's a different issue.

@jordwalke had some helpful pointers - looks like `xterm.js` had some similiar problems (and fixed them). This issue: https://github.com/xtermjs/xterm.js/issues/985 and this PR: https://github.com/xtermjs/xterm.js/issues/985 are especially relevant.

They ensured that their text are aligned with _pixel_ boundaries. This would be a good next-fix for us (#2090 ensures the text is aligned with the 'y' pixel position, but we still need to align with the 'x' pixel position - this is a bit more work because it impacts ligatures and the way we render tokens), if it turns out the WebGL strategy isn't feasible for some reason.

@dw9694 - have you tried the latest master build, with #2090 ? Would be worth seeing if that helps your case.

One other piece of diagnostic information that would be helpful is the result of running these commands in the console (`Control+Shift+P` -> `Debug: Open Developer Tools`), and then:
```
Oni.editors.activeEditor._neovimEditor._store.getState().fontPixelWidth
Oni.editors.activeEditor._neovimEditor._store.getState().fontPixelHeight
```

That can help us figure out if there is a fractional size. If so, we can try small tweaks to the fontSize (like `14.1px` or `13.9px` to see if we can get something closer to whole cell sizes).

**Comment by dw9694** [comment]
> @dw9694 - have you tried the latest master build, with #2090 ?

Yes, I installed the latest version. No result.
![2018-04-17 05-32-48](https://user-images.githubusercontent.com/6016334/38845849-181fddee-4202-11e8-824b-c11171f1f45f.png)

**Comment by cormacrelf** [comment]
@bryphe - Same here on master. `ui.fontSmoothing` didn't make any difference. It also appears that the text at the top is sharp, but further down, gets blurry again. I think it gets sharper at the bottom. This is anchored to the whole Oni window, not one particular editor window.

![oni-blurry-text-rendering-2f80093](https://user-images.githubusercontent.com/378760/38886138-706158b4-42b8-11e8-8668-1a6d83d8c98c.png)

**Comment by Cryza** [comment]
I agree that these issues are most probably caused by bad antialiasing. Due to the way modern displays arrange their subpixels, this is somewhat different between the x and the y direction.

For the vertial antialiasing that this issue is about, unfortunately subpixel antialiasing won't help us. That means we have to ensure that the lines don't need to be antialiased too much in the first place. @bryphe already added some logic to ensure that our lines always align to full pixel boundaries, which should ensure that most fonts render without too much vertical antialiasing (most monospace fonts are optimized for integer font sizes)

For the x direction, the problem should in fact only be present if we do not have subpixel-antialiasing. That technique essentially increases the horizontal resolution by a *theoretical* factor of 3. That's why with subpixel-antialiasing enabled, the canvas-renderer normally creates nice results without having to align the characters to full pixel boundaries. In our WIP WebGL renderer, we will have to put a little more effort into the horizontal placement of letters to fully make use of the subpixel antialiasing, but it should be possible for us to create a result that is indistinguishable from the canvas renderer.

**Comment by Cryza** [comment]
Some more ideas I had:

We calculate the line height based on the measured pixel height of the character "H" in the given font  by default. Since most of the issues seem to have happened on platforms different from Windows and Mac OS, I had a thoguht:
Maybe the default fonts for your platforms render that character to a non-integer height and thus create vertical antialiasing?

It would in any case be great to know the exact height that the non-canvas cursor element is rendered at, since it uses the same height. If you open the electron devtools and search for the first element in the DOM that has `translateZ(0)` as an inline style, you should be able to find the cursor.

For me, it renders to 16px in height as can be seen here:
![screen shot 2018-04-24 at 21 52 45](https://user-images.githubusercontent.com/4516798/39210838-29111434-480a-11e8-838d-1255e736ffa2.png)

Can you please check if it is an integer value for you as well?

**Comment by CrossR** [comment]
I can't speak for everyone but I'm getting this somewhat on Windows 10, with Fira Code. (4K + 150% scaling so that won't help).

After chatting with Bryphe about it, he recommended calling `Oni.editors.activeEditor._neovimEditor._store.getState().fontPixelHeight` in the dev console, which for me across 2 devices gave `16.666666984558105`.

There was also a bit of discussion about using the letter `H` since most of my issues (not related to fuzzy-ness but subpixel antialiasing crack though using your experimental branch seems to have fixed that) are with brackets and braces, which are seemingly taller.

**Comment by bryphe** [comment]
@Cryza - good timing, we were just discussing it 😄  Yes, I was wondering if perhaps we should align our height measurement to pixel boundaries? I do think this is definitely a contributing factor to some of the cases where we see blurry lines of text.

We measure the text right now here:
https://github.com/onivim/oni/blob/479fd42a4912be847d69297f9ff81267804f160a/browser/src/Font.ts#L29

I believe a relatively safe change would be to change this line:
 `const height = rect.height`
to:
`const height = Math.ceil(rect.height)`

In general, there is already padding between the lines by default, so it wouldn't be very noticeable - and it would help us in aligning the text to pixel boundaries. I believe this would be a relatively safe change. In addition, it seems that if `editor.linePadding` is not a multiple of 2, it could cause a subpixel offset due to this code: 
https://github.com/onivim/oni/blob/3c13ad2038e59c8e6529c6e0ad5f26e4222a652f/browser/src/Renderer/CanvasRenderer.ts#L330

(In other words, if you set `editor.linePadding` to 1, we'd draw your text at 0.5 pixel offsets, which could be a contributing factor). We may want to restrict that to be an integer size as well - with those two changes, we'll always draw rows aligned to an integer pixel coordinate.
